src/Language/PureScript/TypeChecker/Unify.hs

```diff
@@ -195,13 +195,31 @@ replaceTypeWildcards = everywhereOnTypesM replace
 -- |
 -- Replace outermost unsolved unification variables with named type variables
 --
-varIfUnknown :: [(Unknown, SourceType)] -> SourceType -> SourceType
-varIfUnknown unks ty =
-  mkForAll (toBinding <$> unks) $ go ty
+varIfUnknown :: forall m. (MonadState CheckState m) => [(Unknown, SourceType)] -> SourceType -> m SourceType
+varIfUnknown unks ty = do
+  bn' <- traverse toBinding unks
+  ty' <- go ty
+  pure $ mkForAll bn' ty'
   where
-  toName = T.cons 't' . T.pack .  show
-  toBinding (a, k) = (getAnnForType ty, (toName a, Just $ go k))
-  go = everywhereOnTypes $ \case
-    (TUnknown ann u)
-      | Just _ <- lookup u unks -> TypeVar ann (toName u)
-    t -> t
+  toName = T.cons 't' . T.pack . show
+
+  toBinding :: (Unknown, SourceType) -> m (SourceAnn, (T.Text, Maybe SourceType))
+  toBinding (a, k) = do
+    mn <- lookupUnkName a
+    k' <- go k
+    case mn of
+      Just n ->
+        pure (getAnnForType ty, (n <> T.pack (show a), Just k'))
+      Nothing ->
+        pure (getAnnForType ty, (toName a, Just k'))
```

Comment:
Is it ever possible to hit this Nothing case? In other words, should this throw an `internalError` instead?

Reply:
Untyped declarations would still trigger this e.g.
```hs
-- forall t1. t1 -> t1
id a = a
```

Reply:
Ah, gotcha.